src/coreclr/tools/Common/TypeSystem/Common/MetadataVirtualMethodAlgorithm.cs

```diff
@@ -975,34 +959,54 @@ public static MethodDesc ResolveVariantInterfaceMethodToStaticVirtualMethodOnTyp
             return null;
         }
 
+        private static MethodDesc TryResolveInterfaceMethodOnVariantCompatibleInterface(TypeDesc compatibleInterfaceType, MethodDesc interfaceMethod)
+        {
+            Debug.Assert(compatibleInterfaceType.CanCastTo(interfaceMethod.OwningType));
+            if (compatibleInterfaceType is MetadataType mdType)
+            {
+                foreach (MethodDesc runtimeInterfaceMethod in mdType.GetVirtualMethods())
+                {
+                    if (runtimeInterfaceMethod.Name == interfaceMethod.Name &&
+                        runtimeInterfaceMethod.Signature == interfaceMethod.Signature)
+                    {
+                        return runtimeInterfaceMethod;
+                    }
+                }
+            }
+            return null;
+        }
+
         /// <summary>
         /// Try to resolve a given virtual static interface method on a given constrained type and return the resolved method or null when not found.
         /// </summary>
         /// <param name="constrainedType">Type to attempt method resolution on</param>
+        /// <param name="interfaceType">Interface declaring the method</param>
         /// <param name="interfaceMethod">Method to resolve</param>
         /// <returns>MethodDesc of the resolved method or null when not found (runtime lookup must be used)</returns>
-        private static MethodDesc TryResolveVirtualStaticMethodOnThisType(MetadataType constrainedType, MethodDesc interfaceMethod)
+        private static MethodDesc TryResolveVirtualStaticMethodOnThisType(TypeDesc constrainedType, MethodDesc interfaceMethod)
```

Comment:
Why make this take TypeDesc and then not actually handle anything but MetadataType?